### ISSUE TITLE
Timeline Logging Duplicates of Operations

### ISSUE THREAD

**Issue by xilenteyex** [tensorflow/tensorflow#17910]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**: 1.6.0-rc1
- **Python version**: 2.7
- **Bazel version (if compiling from source)**: 0.11.0
- **GCC/Compiler version (if compiling from source)**: 5.4
- **CUDA/cuDNN version**: 9.1/7.0
- **GPU model and memory**: Tesla k80 (11441MiB)
- **Exact command to reproduce**: python cifar10_train.py

### Describe the problem
I used timeline to profile the time taken by each operation of the standard cifar10 model available in tensorflow/models repo. After looking at the logfile, it looks like logs of some of the operations are duplicated i.e. it looks like some of the operations in the graph are ran multiple times over the single run of the complete graph. For example, Operation "gradients/conv2/Conv2D_grad/Conv2DBackpropFilter" (link to logfile : https://gist.github.com/xilenteyex/d54305e0448e1aa3d878872c45b8ed3a#file-timeline-1-json-L2270) is logged multiple times. Is this some sort of bug or am I missing something? 

Thanks a lot for looking into this issue!  

### Source code / logs
cifar10 example : https://github.com/tensorflow/models/tree/master/tutorials/image/cifar10
here is a the link to my modified version of cifar10_train.py in which I added logging : 
https://gist.github.com/xilenteyex/b6fab3a5abdb65bf674aa7d0a4ec4b5c
one of the sample log files is : https://gist.github.com/xilenteyex/d54305e0448e1aa3d878872c45b8ed3a

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by aselle** [comment]
@prb12, PTAL.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @prb12: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @prb12: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @prb12: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @prb12: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @prb12: It has been 75 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @prb12: It has been 90 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @prb12: It has been 105 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @prb12: It has been 120 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by xilenteyex** [comment]
Any sort of Input is much appreciated. 
Thanks!

**Comment by prb12** [comment]
Well, it's very hard to see what the problem might be from the "modified code" link you sent because the timeline capturing code is commented out in a way that obfuscates the original control-flow:
```
      #run_metadata = tf.RunMetadata()
      #count = 0
      while not mon_sess.should_stop():
#        mon_sess.run(train_op, options=tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE), run_metadata=run_metadata)
        mon_sess.run(train_op)
#        from tensorflow.python.client import timeline
#        trace = timeline.Timeline(step_stats=run_metadata.step_stats)
#        trace_file = open('timeline_%d_cifar10_cpu.ctf.json' % (count), 'w')
#        trace_file.write(trace.generate_chrome_trace_format())
#        count += 1
```

But... the fact that you create the `tf.RunMetadata` protobuf outside the loop and potentially reuse it many times means that you quite probably have merged a bunch of `StepStats` into the same protobuf.  The `Timeline` code would be very confused by this.  

To debug, I would suggest writing `step_stats` to a file, or printing it out.  If there are numerous copies of the same step/ops in there then its not a timeline problem.

But first I would try making a new `tf.RunMetadata` object for each `sess.run` call, make sure you just trace a *single* step and only write the trace file once.

Failing any of those, I would suspect maybe something funny with the CUPTI library version and/or GPU driver-  but that seems very unlikely.

**Comment by xilenteyex** [comment]
Hi @prb12 ,
Thanks for replying.

I am trying to understand logging in tensorflow in depth. I think cifar is slightly a bigger model to start with. Now, I am using a MNIST example. Script that I ran can be found [(here)](https://gist.github.com/xilenteyex/f652be2306573020eec9152a87915324). Following is the list of issues:

- I made sure that run_metadata is called separately for every sess.run call. I am still seeing multiple entries for the same operation in the same thread . For example, op named "train/Adam/update_layer2/weights/Variable/ApplyAdam" is logged three times in pid = '11'. Does this really mean that my model is designed in such a way that it requires "train/Adam/update_layer2/weights/Variable/ApplyAdam" op to be executed three times for each sess.run call ? (step_stats can be found [here](https://gist.github.com/xilenteyex/aef5dd189fe210acd4d8720c804b43a8) and timline can be found [here](https://gist.github.com/xilenteyex/6ac4f81d767bc6a2cab89822a6d38584))

- According to my understanding (please correct me, if I am wrong). There are 4 threads running on GPU, one is responsible for MEMCPYHtoD, one for MEMCPYDtoH, one for MEMCPYPtoP and one thread for execution of compute nodes. But, I see a lot of GPU streams in the timeline, is this just an artifact of timeline to enable it to be visualized in the timeline or it has some other significance ?

- Also, reading ur comments, on [other issues](https://github.com/tensorflow/tensorflow/issues/1824#issuecomment-244251867), I understand that the pid named : "/job:localhost/replica:0/task:0/device:GPU:0 Compute" just logs the time it takes to launch the kernel on the GPU. Actual execution time is logged in the streams named "/device:GPU:0/stream:* Compute". But after looking at the timestamps and the duration, it looks like "/job:localhost/replica:0/task:0/device:GPU:0 Compute" is actually logging the total time (kernel launch + actual compute node execution time on GPU) e.g. 
for the op "train/Adam/update_layer2/weights/Variable/ApplyAdam", there is one entry in "/job:localhost/replica:0/task:0/device:GPU:0 Compute" (pid : 9) which starts at timestamp : 1536530947160930 and ends at 1536530947161023, but there are three entries in the pid : 11, named : "/device:GPU:0/stream:22 Compute". These three entries start and end with timestamps as follows:
(start, end)
(1536530947160982, 1536530947160984)
(1536530947160999, 1536530947161001)
(1536530947161016, 1536530947161021)
All three of these have starting time after the start timestamp of "/job:localhost/replica:0/task:0/device:GPU:0 Compute" and ends before the end timestamp of "/job:localhost/replica:0/task:0/device:GPU:0 Compute". What does this really mean ? What does these gaps in these three timeslots show? 
Also, what is the difference between two GPU streams "/device:GPU:0/stream:22 Compute" (pid: 11) and "/device:GPU:0/stream:all Compute" (pid: 7)

- In pid: "5" named "/device:GPU:0/memcpy Compute", I assumed that it will only contain memcpy nvidia cuda operations. But, there is an op of type "Assign" in this stream. Is this the expected behavior ?

- In order to log the time it takes to send a tensor from one one device to another, is it enough to look at the duration and timestamps of memcpy ops ? or do i need to enable logging of send/recv op nodes using the hack as discussed in this [github issue](https://github.com/tensorflow/tensorflow/issues/4809) ? If those memcpy ops report the correct times and we don't really need the hack to enable the send/recv ops. Is there a way to map MEMCPY ops to tensors ?

- Also, whats the purpose of "/device:GPU:0/stream:* Tensors" and "/job:localhost/replica:0/task:0/device:GPU:0 Tensors" processes ?

- Overall, for a given tensorflow graph, I am interested in creating a timeline programmatically (not UI, raw values), using which I can find the time at which an op (compute as well as communication op) started and the time it took to complete. Do you think tensorflow.timeline is the right way to go about doing this ?

P.S. : I am sorry, I know this is not a discussion forum and lack of documentation about usage and what each field represents about logging, I am unable to figure out how to use this exactly. I tried asking the same thing over stackoverflow a few times, but my questions remained unanswered. I am stuck at this for almost 3 months now. If you can clarify all these things, that wil be awesome! 

Thanks a lot for looking into this.

**Comment by xilenteyex** [comment]
@prb12 ,  it will be great if you can look into this.

Thanks!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @prb12: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @prb12: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @prb12: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mohantym** [comment]
We see that you are using old version of Tensorflow which is officially considered as end of life, We recommend that you upgrade to 2.4 or later version and let us know if the issue still persists in newer versions .Please open a new issue in case you face any errors, we will get you the right help .Thanks!
